src/jsx.ts

```diff
@@ -0,0 +1,39 @@
+import {vnode, VNode, VNodeData} from './vnode';
+import {h, VNodeChildren} from './h';
+
+export type FunctionComponent = (props: {[prop: string]: any} | null, children?: VNode[]) => VNode;
+
+function flattenAndFilter(children: VNodeChildren[], flattened: VNode[]): VNode[] {
+  for (const child of children) {
+    // filter out falsey children, except 0 since zero can be a valid value e.g inside a chart
+    if (child !== undefined && child !== null && child !== false &&  child !== '') {
+      if (Array.isArray(child)) {
+        flattenAndFilter(child, flattened);
+      } else if (typeof child === 'string' || typeof child === 'number' || typeof child === 'boolean') {
+        flattened.push(vnode(undefined, undefined, undefined, String(child), undefined));
+      } else{
+        flattened.push(child);
+      }
+    }
+  }
+  return flattened;
+}
+
+/**
+ * jsx/tsx compatible factory function
+ * see: https://www.typescriptlang.org/docs/handbook/jsx.html#factory-functions
+ */
+export function jsx(tag: string | FunctionComponent, data: VNodeData | null, ...children: VNodeChildren[]): VNode {
+  const flatChildren = flattenAndFilter(children, []);
+  if (typeof tag === 'function') {
+    // tag is a function component
+    return tag(data, flatChildren);
+  } else {
+    if (flatChildren.length == 1 && !flatChildren[0].sel && flatChildren[0].text) {
+      // only child is a simple text node, pass as text for a simpler vtree
+      return h(tag, data, flatChildren[0].text);
```

Comment:
Is there a test covering this, please?

Reply:
yeah that's already covered by the first test.

https://github.com/snabbdom/snabbdom/blob/8624fe5d0b574a7dc842e7be7d718548234a54b1/test/jsx.tsx#L6-L15

and the else part is covered by the flattens test.

https://github.com/snabbdom/snabbdom/blob/8624fe5d0b574a7dc842e7be7d718548234a54b1/test/jsx.tsx#L16-L35

Reply:
Ideally, the tests should convey their coverage to the reader. If a future author edits this test he might alter it in a way that reduces coverage.